# 2. scikit-learnで統計的学習

## 2.1 はじめに

本ノートブックでは，[A tutorial on statistical-learning for scientific data processing](http://scikit-learn.org/stable/tutorial/statistical_inference/index.html)を参考に，scikit-learnにおける統計的学習について復習します．

統計的学習とは，[統計的推論](https://kotobank.jp/word/%E7%B5%B1%E8%A8%88%E7%9A%84%E6%8E%A8%E8%AB%96-103447)を目的とした，機械学習の応用の一つです．

## 2.2 下準備

### Datasets

scikit-learnでは，２次元配列でdatasetを表現します．一次元目を**samples** axis，二次元目を**features** axisと呼びます．

以下は，iris datasetの例です．

In [1]:
from sklearn import datasets
iris = datasets.load_iris()
data = iris.data
data.shape

(150, 4)

それぞれ4つのfeature（萼片長，萼片幅，花弁長，花弁幅）を持つsampleが150個あることを示しています．詳細は，`iris.DESCR`をご参照ください．

In [2]:
iris.DESCR

'Iris Plants Database\n====================\n\nNotes\n-----\nData Set Characteristics:\n    :Number of Instances: 150 (50 in each of three classes)\n    :Number of Attributes: 4 numeric, predictive attributes and the class\n    :Attribute Information:\n        - sepal length in cm\n        - sepal width in cm\n        - petal length in cm\n        - petal width in cm\n        - class:\n                - Iris-Setosa\n                - Iris-Versicolour\n                - Iris-Virginica\n    :Summary Statistics:\n\n    ============== ==== ==== ======= ===== ====================\n                    Min  Max   Mean    SD   Class Correlation\n    ============== ==== ==== ======= ===== ====================\n    sepal length:   4.3  7.9   5.84   0.83    0.7826\n    sepal width:    2.0  4.4   3.05   0.43   -0.4194\n    petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)\n    petal width:    0.1  2.5   1.20  0.76     0.9565  (high!)\n    ============== ==== ==== ======= ===== ============

外部データに対しscikit-learnを使うためには，形状を`(n_samples, n_features)`に整える必要があります．

### 予測器（Estimator）

**フィッティング**：scikit-learnの主なAPIは，予測器をデータにフィッティングさせるためのものです．予測器は，データから学習するあらゆる問題（分類，回帰，クラスタリング，次元削減）に対応しています．

全ての予測器に対し，`fit`メソッドを介してdatasetを入力できます．

In [8]:
from sklearn import svm
clf = svm.SVC()
clf.fit(iris.data[:-1], iris.target[:-1])

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

**Parameter**：予測器の全てのparameterは，初期化時に設定できますし，変数として直接設定することもできます．

In [10]:
clf2 = svm.SVC(gamma=0.001)

In [11]:
clf2.gamma=0.01

**Estimated parameter**：予測器は，学習により幾つかのパラメータを予測します．これらはアンダーバー（`*_`）で終わるattributeとして，ユーザに提供されます．

In [9]:
clf.support_

array([ 13,  15,  18,  23,  24,  41,  44,  50,  52,  54,  56,  57,  60,
        63,  66,  68,  70,  72,  76,  77,  78,  83,  84,  86,  98, 100,
       106, 110, 118, 119, 121, 123, 126, 127, 129, 131, 133, 134, 138,
       141, 142, 146, 147])

## 2.3 教師あり学習

## 2.4 モデル選択

## 2.5 教師なし学習

## 2.6 まとめ

## 2.7 ヘルプ